src/gt4py/next/ffront/decorator.py

```diff
@@ -124,18 +120,34 @@ def from_function(
         backend: Optional[next_backend],
         grid_type: Optional[GridType] = None,
     ) -> Program:
-        source_def = SourceDefinition.from_function(definition)
-        closure_vars = get_closure_vars_from_function(definition)
-        annotations = typing.get_type_hints(definition)
-        past_node = ProgramParser.apply(source_def, closure_vars, annotations)
+        program_def = stages.ProgramDefinition(definition=definition, grid_type=grid_type)
         return cls(
-            past_node=past_node,
-            closure_vars=closure_vars,
+            definition_stage=program_def,
             backend=backend,
-            definition=definition,
-            grid_type=grid_type,
         )
 
+    @property
+    def definition(self):
+        return self.definition_stage.definition
+
+    @property
+    def grid_type(self):
+        return self.definition_stage.grid_type
+
+    @functools.cached_property
+    def past_stage(self):
+        if self.backend is not None and self.backend.transformer is not None:
+            return self.backend.transformer.func_to_past(self.definition_stage)
+        return next_backend.DEFAULT_TRANSFORMS.func_to_past(self.definition_stage)
+
+    @property
+    def past_node(self):
+        return self.past_stage.past_node
+
+    @property
+    def closure_vars(self):
+        return self.past_stage.closure_vars
```

Comment:
Same, probably not needed.

Reply:
Some of this stuff might be used in icon4py. I would suggest adding a todo comment and a deprecation warning.